src/Infrastructure/Adapter/APIAdresseGeocoder.php

```diff
@@ -0,0 +1,114 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Infrastructure\Adapter;
+
+use App\Application\Exception\GeocodingFailureException;
+use App\Application\GeocoderInterface;
+use App\Domain\Geography\Coordinates;
+use Symfony\Contracts\HttpClient\Exception\DecodingExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+final class APIAdresseGeocoder implements GeocoderInterface
+{
+    public function __construct(
+        private HttpClientInterface $http,
+    ) {
+    }
+
+    public function computeCoordinates(string $address, ?string $postalCodeHint = null): Coordinates
+    {
+        // See: https://adresse.data.gouv.fr/api-doc/adresse
+        $url = 'https://api-adresse.data.gouv.fr/search/';
```

Comment:
Je pense qu'il serait préférable de mettre les URLs d'API dans le fichier de configuration.

Reply:
Dans quel fichier est-ce qu'on mettrait ça ? Le `.env` je suppose, comme ça on peut configurer l'URL en local ou en live ? Je vais faire ça.

Reply:
@mmarchois Je viens de réaliser que le `Step2ControllerTest` fait un vrai appel à l'API Adresse, puisqu'on ne mock pas l'URL dans les tests d'intégration.

Dans le cadre de tests d'intégration, ça me semble à moitié problématique, car un test d'inté peut appeler les services externes (on interagit avec une vraie DB), mais il est préférable qu'ils soient contrôlés. Ici, on utilise la vraie API Adresse sur laquelle on n'a pas de contrôle, contrairement à la DB de test...

Comment tu t'y prendrais pour ne pas vraiment appeler l'API Adresse ? Un serveur "mock" local ?

Reply:
Oups je m'apercois que je n'avais pas répondu à ta question. Tu pourrais passer par un mock de l'HTTPClient. Différents exemples existent sur la toile (par ex: https://www.strangebuzz.com/en/blog/simple-api-mocking-with-the-symfony-http-client)

Reply:
On est tombés exactement sur la même ressource :smile: